### PR TITLE
Dependency injection for operations

### PR DESCRIPTION
**Pull request by sebasmagri**
Inject dependencies on operations by the means of binding the operation handlers so that they get `this.dependencies`

### CODE REVIEW

**Comment by sebasmagri** [comment]
@jsdevel sure, was adding some notes but wanted some feedback in advance...

File: packages/openapi-framework/index.ts [hunk 1]
```diff
@@ -260,7 +260,10 @@ export default class OpenAPIFramework implements IOpenAPIFramework {
                 acc[METHOD_ALIASES[method]] = (() => {
                   const innerFunction: any = operation;
                   innerFunction.apiDoc = methodDoc;
-                  return innerFunction;
+                  // Operations get dependencies injected in `this`
+                  return innerFunction.bind({
+                    dependencies: this.dependencies
```
Comment (jsdevel) [inline packages/openapi-framework/index.ts]: also, perhaps we should create a new object, so the operations can't modify the reference?

i.e.

```
  dependencies: {...this.dependencies}
```

File: packages/express-openapi/README.md [hunk 2]
```diff
@@ -743,7 +743,8 @@ app.listen(3000);
 ```
 
 Operations also get `args.dependencies` injected as
-`this.dependencies` on the function scope.
+`this.dependencies` on the function scope. This requires the operation
```
Comment (jsdevel) [inline packages/express-openapi/README.md]: can you reword this to be:

```
**Note:** `args.dependencies` is added to the operation's `this` context via `operation.bind({dependencies: args.dependencies})`.  This requires your operations to be regular [function expressions](https://developer.mozilla.org/en-US/docs/web/JavaScript/Reference/Operators/function) and not [arrow functions](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions) (due to the fact that `this` is lexical to the surrounding scope in an arrow function and cannot be bound to anything else).
```

Comment (sebasmagri) [reply packages/express-openapi/README.md]: Sure!

File: packages/express-openapi/README.md [hunk 3]
```diff
@@ -742,8 +742,42 @@ initialize({
 app.listen(3000);
 ```
 
-**Note:** Handlers in args.paths take precedence over handlers in args.operations for
-historical reasons.
+Operations also get `args.dependencies` injected as
```
Comment (jsdevel) [inline packages/express-openapi/README.md]: Same here

File: packages/express-openapi/README.md [hunk 4]
```diff
@@ -742,8 +742,47 @@ initialize({
...
+      this.log('calling request handler');
+      res.send('foo');
+    }
+  }
+});
+
+app.listen(3000);
+```
+
+**Note:** `args.dependencies` is added to the operation's `this`
```
Comment (jsdevel) [inline packages/express-openapi/README.md]: Sorry, i thought this section was for args.paths.  You can remove this Note, and the one in args.paths if the one in args.paths is no longer valid.

**Comment by sebasmagri** [comment]
@jsdevel Hi! I've applied the requested changes in docs. Anything else you think we should consider?

File: packages/express-openapi/README.md [hunk 5]
```diff
@@ -742,8 +742,47 @@ initialize({
...
+
+initialize({
+  app,
+  apiDoc: './apiDoc.yml',
+  dependencies: {
+    log: console.log
+  },
+  operations: {
+    getFoo: function(req, res) {
+      this.log('calling request handler');
```
Comment (jsdevel) [inline packages/express-openapi/README.md]: should this be `this.dependencies.log`?  If not, then the bind example in the note above should be `.bind({...args.dependencies`

**Comment by sebasmagri** [comment]
> sorry to have so much back and forth over this. please look at the file as a whole and determine what makes sense. i'm eager to merge this!

You're right! I'm gotta check it out.

**Comment by sebasmagri** [comment]
@jsdevel I think it's much better now.

**Comment by jsdevel** [comment]
thanks @sebasmagri !  is the note about operations taking precedence in args.paths still valid?

**Comment by sebasmagri** [comment]
@jsdevel it was not anymore as per [comment](https://github.com/kogosoftwarellc/open-api/pull/353#issuecomment-462850271). I updated the readme to reflect it.

**Comment by jsdevel** [comment]
You killed it @sebasmagri!  Thanks!

**Comment by sebasmagri** [comment]
@jsdevel great! thanks a lot... looking forward to have this in a release ;-)

**Comment by jsdevel** [comment]
published in the latest version @sebasmagri

**Comment by sebasmagri** [comment]
Thanks @jsdevel
